app/src/backlight.c

```diff
@@ -0,0 +1,234 @@
+
+static int zmk_backlight_update_brt() {
+    uint8_t percent = state.brightness * 100 / (BACKLIGHT_LEVELS - 1);
+    return led_set_brightness(led_bl, 0, percent);
+}
+
+static int zmk_backlight_update() {
+    int rc = zmk_backlight_update_on();
+    if (rc != 0) {
+        return rc;
+    }
+
+    rc = zmk_backlight_update_brt();
+    if (rc != 0) {
+        return rc;
+    }
+
+    return 0;
+}
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+static int bl_settings_set(const char *name, size_t len, settings_read_cb read_cb, void *cb_arg) {
+    const char *next;
+
+    if (settings_name_steq(name, "state", &next) && !next) {
+        if (len != sizeof(state)) {
+            return -EINVAL;
+        }
+
+        int rc = read_cb(cb_arg, &state, sizeof(state));
+        if (rc < 0) {
+            return rc;
+        }
+
+        return zmk_backlight_update();
+    }
+
+    return -ENOENT;
+}
+
+static struct settings_handler bl_conf = {.name = "backlight", .h_set = bl_settings_set};
+
+static void zmk_backlight_save_state_work() {
+    settings_save_one("backlight/state", &state, sizeof(state));
+}
+
+static struct k_delayed_work backlight_save_work;
+#endif // IS_ENABLED(CONFIG_SETTINGS)
+
+static int zmk_backlight_save_state() {
+#if IS_ENABLED(CONFIG_SETTINGS)
+    k_delayed_work_cancel(&backlight_save_work);
+    return k_delayed_work_submit(&backlight_save_work, K_MSEC(CONFIG_ZMK_SETTINGS_SAVE_DEBOUNCE));
+#else
+    return 0;
+#endif
+}
+
+static int zmk_backlight_init(const struct device *_arg) {
+    led_bl = device_get_binding(BACKLIGHT_LABEL);
+    if (led_bl) {
+        LOG_INF("Found LED backlight device %s", BACKLIGHT_LABEL);
+    } else {
+        LOG_ERR("LED backlight device %s not found", BACKLIGHT_LABEL);
+        return -EINVAL;
+    }
+
+    state = (struct backlight_state){
+        brightness : CONFIG_ZMK_BACKLIGHT_LEVEL_START - 1,
+        on : IS_ENABLED(ZMK_BACKLIGHT_ON_START)
+    };
+
+#if IS_ENABLED(CONFIG_SETTINGS)
+    settings_subsys_init();
+
+    int err = settings_register(&bl_conf);
+    if (err) {
+        LOG_ERR("Failed to register the backlight settings handler (err %d)", err);
+        return err;
+    }
+
+    k_delayed_work_init(&backlight_save_work, zmk_backlight_save_state_work);
+
+    settings_load_subtree("backlight");
+#endif
+
+    return zmk_backlight_update();
+}
+
+static int zmk_backlight_set_on(bool on) {
+    if (!led_bl) {
+        return -ENODEV;
+    }
+
+    state.on = on;
+
+    int rc = zmk_backlight_update_on(led_bl, 0);
+    if (rc != 0) {
+        return rc;
+    }
+
+    return zmk_backlight_save_state();
+}
+
+static int zmk_backlight_change_brt(int direction) {
+    if (!led_bl) {
+        return -ENODEV;
+    }
+
+    int b = state.brightness + direction;
+    if (b < 0) {
+        b = 0;
+    } else if (b >= BACKLIGHT_LEVELS) {
+        b = BACKLIGHT_LEVELS - 1;
+    }
+
+    // turn on backlight when change brightness
+    state.on = true;
```

Comment:
```suggestion
    state.on = b != 0;
```
This would make it so that when brightness is 0 we turn off the PWM and save power.